CIPs/0005.md

```diff
@@ -0,0 +1,68 @@
+
+The goals for this CIP are the following:
+
+1) Minimize bandwidth requirements for the network's full nodes.  Some of the potential full node operators need to use ISPs that have expensive networking rates.  We need to ensure that adoption of the full nodes is as broad and diverse as possible, so that the consumer end users will have good access to the blockchain via the mobile light client.
+2) Maintain a healthy overlay network for IBFT.   This is to maximize consistent block times, which obviously has direct impact to the consumer end users.
+
+This CIP is an iteration of the Announce protocol that is implemented in https://github.com/celo-org/celo-blockchain/pull/816.  The original protocol does optimize for 2), but not for 1).
+
+## Goals
+
+- Reduce bandwidth requirements of the full nodes.
+- Maximize availability of the IBFT overlay network.
+
+## Proposed Solution
+
+The announce protocol has to be able to deal with two fundamental faulty scenarios.  For a given validator pair, when only one of the validators has the correct enodeURL of the other (the asymmetric ignorance case) and when both of the validators don't have the correct enodeURL of the other (the mutual ignorance case).
+
+To deal with these cases, this proposal proposes adding the following high level components to the existing Announce protocol.  1) Add the capability for a validator to directly communicate to another remote validator it's assigned enodeURL, 2) add fast discovery of validator enode mapping changes, and 3) add capability for a validator to query another validator for it's assigned enodeURL via a gossiped message.  Details of these components are in the subsections below.
+
+The asymmetric ignorance case can then be handled by having the validators directly communicate to each remote validator it's assigned enodeURL whenever the enodeURL mapping is updated.  Note that this will need to happen for all remote validators, even if the assigned URL is not updated for a subset of the validators, since all validators will need to know what is the updated version number.  Also, this direct communication will need to be handled when there is and is not an existing connection between validators.
+
+The mutual ignorance case can be handled by having each validator monitor when a validator enode mapping changes, and when it realizes that it's changed, it will gossip a query message to the validator for the updated assigned enodeURL.  That validator will then directly communicate it to the querying validator.
+
+### Direct communication of assigned enodeURL
+
+This component will include the following changes:
+
+1)  Announce message format will need to be changed so that only one authenticated and encrypted enodeURL (along with the version number) can be put in the message payload
```

Comment:
Rather than describing the difference from the current implementation, I think this CIP would be clearer with explicit descriptions of the intended protocols.

Reply:
Agreed

Reply:
You're right that it doesn't need to be encrypted.  My thought is that it'll be easier to code by having one format for the direct announce and gossiped announce.

Reply:
> Rather than describing the difference from the current implementation, I think this CIP would be clearer with explicit descriptions of the intended protocols.

As the messages are added within Geth, I'll update this CIP to include those message formats.